Add: 무기 기본 및 발사체 기본 코드 구현. GameplayAbility, GameplayEffect, AttributeSet, Excution, BP(WeaponBase, WeaponRifle, ProjectileBase) 추가

wjddnjs6520 opened the pull request:
## Add
### ShooterWeaponBase 클래스
- 무기 베이스 클래스
- Mesh(외형)와 Muzzlelocaion(발사 위치) 존재
- GA Class
  - 블루프린트로 만든 후 GA 할당
  - WeaponFire: 총 발사
  - WeaponReload: 총 재장전
  - EquipWeapon: 총 장착
    - 해당 어빌리티는 캐릭터의 StartSet에 할당 후 자동 실행되는 어빌리티
 
###  WeaponAttributeSet 추가
- 무기 GAS AttributeSet
-  CurrentAmmo와 MaxAmmo 존재
- 초기 값
  - MaxAmmo:30
  - CurrentAmmo: GetMaxAmmo()

### Gameplay Tags 추가
- InputTag: 어빌리티를 실행할 인풋 태그
  - InputTag_Weapon_Fire
  - InputTag_Weapon_Reload
 - AbilityTag: 어빌리티들에 할당할 태그
   - Ability_Weapon_Equip
   - Ability_Weapon_Fire
   - Ability_Weapon_Reload

### Projectile
- ProjectileMovementComponent를 사용하는 발사체 클래스
- FireInDirection 함수로 발사 방향 지정
- OnHit 시 블루프린트로 구현한 OnProjectileHit 이벤트 실행
- OnProjectileHit는 GA에서 필요한 형태로 구현
 
### Blueprint
- 아래 있는 항목들은 Content/Weapon에 위치해 있습니다
- GA
  - WeaponFire
    - GE_ConsumeAmmo 후 선택한 방법에 따라 Hit된 액터가 ShooterBaseCharacter일 경우 데미지를 줌.
    - ActiveAbility를 LineTrace나 Projectile에 묶인 곳 제일 왼쪽 Cast to ShooterBaseCharacter에 연결 후 실행.
  - WeaponReload
    - 무기의 CuurentAmmo를 MaxAmmo로 변경
  -  EquipWeapon
    - 장착한 캐릭터에게 무기의 어빌리티 할당
- GE: Excution을 할당 후 GA에서 사용
  - ConsumeAmmo
  - ReloadAmmo
  - TakeDamage
- BP
  - ShooterWeaponBase 
    - Base
    - Rifle
  - WeaponAttributeSet
    - Base
- Excution: AttributeSet의 계산을 맡는 Cpp 기반 블루프린트
  - Consume
    - CurrentAmmo - 1
  - Reload
    - CurrentAmmo = MaxAmmo
  - Damage
    - CurrentHealth(Target) - AttackPower(Activator)
- Projectile: 발사체
  - Base

wjddnjs6520 committed:
Add: ShooterWeaponBase, WeaponAttributeSet 클래스 및 무기 관련 태그 추가

ShooterWeaponBase 클래스 작성 (무기 기본 베이스)
WeaponAttributeSet 클래스 추가 (GAS용 무기 속성 관리)
Gameplay Tag 정의 추가
  - InputTag_Weapon_Fire, InputTag_Weapon_Reload(발사, 장전 입력 태그)
  - Ability_Weapon_Equip, Ability_Weapon_Fire, Ability_Weapon_Reload(장비 장착, 발사, 장전 어빌리티 태그)

wjddnjs6520 committed:
Add: ConsumeAmmo, Reload, Damage Execution 추가

ConsumAmmoExecution: GA_WeaponFire 시 무기의 CurrentAmmo의 값 1 감소
ReloadExecution: GA_WeaponReload 시 무기의 CurrentAmmo의 값 MaxAmmo로 변경
DamageExecution: GA_WeaponFire 시 라인트레이스에 BP_ShooterEnemyCharacterTest가 Hit Actor일 시 해당 Actor의 CurrentHealth 값 1 감소

wjddnjs6520 committed:
Add: GA(WeaponFire, WeaponReload, EquipWeapon), GE(ConsumeAmmo, ReloadAmmo, TakeDamage), BP(Base, Rifle), AS_WeaponAttributeSet 추가

GameAbility 블루프린트 추가
WeaponFire: 총알 발사 어빌리티. 무기의 CurrentAmmo 1 감소. 라인트레이스에 적 액터 Hit 시 적 CurrentHealth 1 감소
WeaponReload: 총알 장전 어빌리티. 무기의 CurrentAmmo를 MaxAmmo로 변경
EquipWeapon: 캐릭터의 Skeletal Mesh에 무기 부착. WeaponFire, WeaponReload 어빌리티 및 WeaponAttributeSet 을 캐릭터의 ShooterAbilitySytstemComponent에 할당

GameplayEffect
ConsumeAmmo: WeaponAttributeSet의  CurrentAmmo 1 감소
ReloadAmmo: WeaponAttributeSet의  CurrentAmmo값 MaxAmmo로 변경
TakeDamage: 라인트레이스에 Hit된 액터의 AttributeSet의 CurrentHealth 1 감소

Blueprint
ShooterWeaponBase: Base 코드의 BP 형태
ShooterWeaponRifle: BP_ShooterWeaponBase를 상속받은 Rifle 무기 블루프린트

AS_WeaponAttributeSet: 무기용 GAS AttributeSet. CurrentAmmo, MaxAmmo 포함

각 기능의 기본 기능만 구현해 놓았습니다.

wjddnjs6520 committed:
Add: ShooterProjectileBase 클래스, BP 생성

ShooterProjectileBase: 발사체 베이스 클래스
스피어 콜리전을 가지고 있고 ProjectileMovementComponent를 통해 발사체를 이동시킵니다.
FireInDirection을 통해 발사체의 이동 방향을 결정할 수 있습니다.
OnHit이 발생하면 OnProjetileHit 이벤트가 발생합니다.
OnProjectileHit는 블루프린트에서 GA마다 다르게 구성 가능합니다.

wjddnjs6520 committed:
Update: GA_WeaponFire - Projectile 방식 추가. GA_WeaponReload - 캐스트 변경

GA_WeaponFire
기존 라인트레이스 방식 외에 Projectile을 이용한 무기 공격 기능을 추가했습니다.
LineTrace와 Projectile로 묶여있는 노드들 맨 왼쪽 첫번째 cast to shooterbasecharacter에 연결 후 GA 실행 가능합니다.
GA_WeaponReload
캐스트를 BP_shootercharacter에서 shooterbasecharacter로 변경하였습니다.

이상한 주석 변경

wjddnjs6520 committed:
refactor: ShooterWeaponBase 코드 변경

GA에서 처리하여 사용하지 않는 변수 및 함수 제거, 함수 로직 정리

hwarang98 commented on Source/Shooter/Public/Items/Weapons/ShooterWeaponBase.h:
```diff
@@ -1,26 +1,52 @@
-	virtual void BeginPlay() override;
+	// 선택 가능한 AttributeSet 클래스 (블루프린트에서 지정)
+	UPROPERTY(EditAnywhere, BlueprintReadOnly, Category = "Weapon|GAS")
+	TSubclassOf<UAttributeSet> WeaponAttributeSetClass;
```
UAttributeSet 해당 클래스도 전방선언해주셔야할것같습니다!

hwarang98 commented on Source/Shooter/Private/Items/Weapons/ShooterWeaponBase.cpp:
```diff
@@ -1,27 +1,59 @@
+				FGameplayAbilitySpec Spec(AbilityClass, 1);
+
+				// InputTag를 Spec의 DynamicAbilityTags에 추가
+				Spec.DynamicAbilityTags.AddTag(InputTag);
```
DynamicAbilityTags 해당 메서드는 언리얼쪽에서 더이상 사용안하는코드라 GetDynamicSpecSourceTags 이걸로 변경해주셔야합니다!

hwarang98 commented on Source/Shooter/Private/Items/Weapons/DamageExecution.cpp:
```diff
@@ -0,0 +1,61 @@
+        return;
+    }
+
+    UAbilitySystemComponent* ActivatorASC = nullptr;
```
동작은 이상없지만 ActivatorASC는 선언된 이후, 실제 사용되기 전에 반드시 값이 할당되므로 초기화할 필요는 없습니다!

hwarang98 commented:
헤더 파일에 public 되어있는것도 한번 봐주시고 private or protected로 보낼수있는건 보내주시면 감사하겠습니다! @wjddnjs6520

hwarang98 commented on Source/Shooter/Public/Items/Weapons/ShooterWeaponBase.h:
```diff
@@ -1,26 +1,52 @@
+	TSubclassOf<UShooterGameplayAbility> ReloadAbilityClass;
+
+	// 장착 어빌리티 클래스
+	UPROPERTY(EditAnywhere, BlueprintReadOnly, Category = "Abilities")
```
이부분도 Abilities를 각각 하나의 변수에서 관리중인데 map이나 array로 관리하면 확장성 면에서 좋을것같아요

`TArray<TSubclassOf<UShooterGameplayAbility>> WeaponAbilities;` 이런식으로 관리해서 함수내부에서 체크하는것도 좋은 방법일것 같습니다!

wjddnjs6520 committed:
Refactor: 무기 관련 코드 수정

ShooterWeaponBase
- AbilityClass 설정을 각 어빌리티마다가 아니라 BP에서 Map을 통해 GA와 InputTag를 설정할 수 있도록 변경
- Ability를 for문을 통해 할당하도록 GiveAbilityToOwner 함수 수정
- 변수와 함수 접근제어자 재지정
- 전방선언과 헤더 파일 확인 후 필요한 것 수정

DamageExecution
- ActivatorASC를 찾는 과정에서 불필요한 nullptr 삭제

Bathboomsoso commented:
고생하셨습니다!!
